### ISSUE TITLE
Set and Map implementations break Safari structuredClone

### ISSUE THREAD

**Issue by SKaistrenko** [immerjs/immer#1149]
## 🐛 Bug Report

If your bug report involves classes, check the [docs](https://immerjs.github.io/immer/docs/complex-objects) on how classes are handled first. For patches, don't file issues if they are not optimal, only if they are incorrect.

## Link to repro

_A bug report without a reproduction is not a bug report. Failing to follow this templately is likely to result in an immediate close & lock of the issue._

https://codesandbox.io/p/sandbox/immer-sandbox-forked-nf577m?file=%2Fsrc%2Findex.ts%3A6%2C8&workspaceId=76a181a1-ec65-4f80-815c-e379a177be13

Unfortunately CodeSandbox doesn't support Safari well - you'll need to run manually on Safari.
On the link above you can see code work ok in Chrome

## To Reproduce

Steps to reproduce the behaviour:
**In Safari browser**
Call `structuredClone` with either a Set or a Map passed through immer

The above op is called within `indexedDB` put operation ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/put))

## Observed behaviour

`DataCloneError: The object can not be cloned., value: [object Object]`

## Expected behaviour

No error - as with plain Map and Set that work well.

## Environment

Latest immer, Safari

- **Immer version:**
- [x] I filed this report against the _latest_ version of Immer
- [x] Occurs with `setUseProxies(true)`
- [x] Occurs with `setUseProxies(false)` (ES5 only)

**Comment by oriSomething** [comment]
It's not related to Safari. `Proxy` seems not be possible to be clone with `structuredClone` 

I've tried running this in Chrome which result in the same error:
```js
x = new Map([[1,1]]); structuredClone(new Proxy(x, {})) === x;
// throws: Uncaught DataCloneError: Failed to execute 'structuredClone' on 'Window': #<Map> could not be cloned.
```

**Comment by mweststrate** [comment]
This looks to me as working as intended, drafts are very special objects
and should never leave the produce function by any other means than
returning it. Why are you structuralCloning those? The output of immer is
immutable anyway, so that seems completely unnecessary?

**Comment by mweststrate** [comment]
Note that there is also current if really needed if you want to leak a
snapshot, although using it is usually a code smell imho
https://immerjs.github.io/immer/current

**Comment by SKaistrenko** [comment]
TLDR: In my understanding immer modifications to original data are leaking out beyond RTK/produce.

We are not trying to run `structuredClone` within produce - this just seems the most straightforward repro.
We use Redux RTK and some of the data we're storing includes `Set`.
After that we're selecting this data within the app cycle and do a buck-up using indexed DB.

It is my understanding that immer is by reference modifying the underlying objects we're storing, resulting in the described behaviour.

I'm able to confirm the above assumption by exporting the Set in question and checking the `===` before putting into IndexedDB.
Meaning
```
const someSet = new Set();
const initialState = { blah: someSet };
...

const stateData = <selectState>;
someSet === stateData.blah // true
```

If I log the Object.keys of the above set before the RTK slice logic, I get an empty list
In the data I select the above return `[add, delete, <some other two keys immer is overwriting>]

**Comment by markerikson** [comment]
@SKaistrenko my first observation is that you shouldn't be putting Sets into Redux state anyway:

- https://redux.js.org/style-guide/#do-not-put-non-serializable-values-in-state-or-actions

**Comment by mweststrate** [comment]
> It is my understanding that immer is by reference modifying the underlying objects we're storing

That understanding is wrong. Not doing that is kinda the point of the whole library :) draft objects are mutable, yes, but neither the inputs or outputs of produce are. So draft objects (they are temporarily mutable copies) behave entirely different from what you get back from produce.

 So in either case structuredClone seems to be a strange thing to use icm with immer. If you structureClone the output of produce, that is moot since you could just directly have handed out the reference itself since it is immutable. If you do it inside produce, you are leaking objects that are temporal and this is by design a thing that you should't be doing (or needing. Why would you even share a ref to an object in the middle of some update process?). Just as you wouldn't be sharing the input argument of a react useState updater function with anything that outlives that very function.

**Comment by SKaistrenko** [comment]
> @SKaistrenko my first observation is that you shouldn't be putting Sets into Redux state anyway:
> 
> * https://redux.js.org/style-guide/#do-not-put-non-serializable-values-in-state-or-actions

Agreed

> If you do it inside produce, you are leaking objects that are temporal and this is by design a thing that you should't be doing

The only reason I exported that reference is to validate my assumption that immer messed with the original ref - which doesn't sound ideal.

**Comment by mweststrate** [comment]
> which doesn't sound ideal.

luckily that isn't the case :) 

closing issue as it is further unintended usage of Immer drafts as documented.
